Change version to 6.0.alpha1

colemanw opened the pull request:
Overview
----------------------------------------
It's happening!

colemanw committed:
Change version to 6.0.alpha1

ufundo commented:
> It's happening!

Jenkins less enthusiastic...

![image](https://github.com/user-attachments/assets/e993d0eb-10e6-4235-9f39-ac4e4e32f772)


Looks like install crashes quite early:

```
[Setup:info] [Preboot.civi-setup.php] Load minimal (non-DB) services
[Setup:info] [InstallSchema.civi-setup.php] Install database schema
[Setup:info] [InstallSchema.civi-setup.php] Load basic tables
Error: Call to a member function getHelper() on null in Civi::schemaHelper() (line 339 of /home/homer/buildkit/build/build-4/web/sites/all/modules/civicrm/Civi.php).
```

Happy to investigate this further unless you are already into it @colemanw ?

ufundo commented:
Oh it looks like the `5` in `civimix-schema@5` is linked to our core version by this call: https://github.com/civicrm/civicrm-core/blob/726d32a47167de9bc88909bfee59ab6cde119b60/mixin/lib/pathload.index.php#L15

If I read it correctly:
- on any 5.x we are registering directory search for `civimix-schema@5.1` which is satisfied by the declaration `civimix-schema@5` in the file https://github.com/civicrm/civicrm-core/blob/726d32a47167de9bc88909bfee59ab6cde119b60/mixin/lib/civimix-schema/pathload.main.php#L4 - which then satisfies the demand for `civimix-schema@5` in all the extension `.civix.php` file

So when we bump core up to 6.x we need to either:
a) hardcode the version in https://github.com/civicrm/civicrm-core/blob/726d32a47167de9bc88909bfee59ab6cde119b60/mixin/lib/pathload.index.php#L15 to `5.1`
b) bump up all the extensions to use 6 (feels unfeasible with contrib)
c) add search directory for the current version and previous version?

Probably an @totten question!

totten committed:
civimix-schema - Update numbering for civicrm-core@6.x

totten commented:
I think my favorite options are:

* Double-register `civmix-schema` in both 5.x and 6.x sequences.
* Just continue with `civimix-schema` in the 5.x sequence. 

In either scenario, the 5.x sequence can be specified formulaically: `6.x = 5.{83+x}`. (_Example_: `6.2 = 5.85`.) This ensures that  the sequence remains legal without needing us to manually bump numbers.

Since `civimix-schema` is already "reloadable" (i.e. no hard-conflicts over PHP symbols), it should work the same even if it's loaded multiple times.

My first impulse was to go with the double-registration -- because it preserves the API POV that numbers match core. OTOH, it does create more numbers floating around `universe`. Keeping the numbers strictly 5.x is *systemically tighter* (and leaves the window open to introduce `6.x` whenever there's an actual BC break).

On my local, I'm able to do basic install either way. Pushing up a commit which shows both.

colemanw commented:
@totten it feels like the `CiviMix` version is unnecessarily tied to the `CiviCRM` version?
Could they be treated independently? Like how our `API` version is still 4?

totten commented:
@colemanw Yes and no.

* From a 30k ft POV, I like this part of  the analogy: we have `APIv3` and `APIv4`, and we could have `APIv5`. These things coexist regardless of the official CiviCRM version.

    ```
    api/v3/                         (the 3.x series)
    Civi/Api4/                      (the 4.x series -- diff contracts; breaking changes)
    ```

    An analogous play for `civimix-schema` might be this layout:

    ```
    mixin/lib/civimix-schema@5      (the 5.x series)
    mixin/lib/civimix-schema@6      (the 6.x series -- diff contracts; breaking changes)
    ```

* From a 5k ft POV, I think a better analogy would be `ext/recaptcha`, `ext/search_kit`, `ext/legacycustomsearches`, etc:
    * Each sub-component can track its own independent number. (Each one has a different physical spot where you can set the number.)
    * In fact, the content probably doesn't change much month-to-month.
    * We're so accustomed to "edit+go" that we would likely forget to maintain the numbers without some process-enforcement.
    * There is a conceptual boundary around each folder. You could copy these folders between different builds/versions. (For `ext/recaptcha/`, we don't really police contracts well enough to encourage that. But for everything under `mixin/`, it's policed more carefully so that it can be copied around -- and at runtime it uses `version_compare()` to decide a winner. For `version_compare()` to be meaningful, you want more decimals. So from a low-level, it's a bit different from APIv3/APIv4.)

totten committed:
civimix-schema - Split off major-version from core. Keep 5.x.

Before:

* 'mixin/lib/civimix-schema' is numbered `5.x.x`
* `mixin/lib/civimix-schema` is strictly pegged to the 'civicrm-core'

After:

* `mixin/lib/civimix-schema@5` is numbered `5.x.x.`
* `mixin/lib/civimix-schema@5` is partially released from the numbering of `civicrm-core`.

Comments:

The version will still increment automatically (month-to-month; in parallel
with core), but it will stay within 5.x. This is because it's supposed to be
strictly SemVer, and we haven't actually made any major changes to it, so
moving to 6.x would be a big hassle.

totten commented:
Yeah, if we're decoupling the version-sequence, then probably the folder-name should indicate the distinctive major-version. This would be more similar to `api/v3`, `Civi\Api4`, or `mixin/smarty@1`.

ufundo commented:
> for everything under mixin/, it's policed more carefully so that it can be copied around

I _think_ I understand the theory. But in practice, is `civimix-schema` copied around?

It seems to me that the key requirements are:
- the core version can provide the version requested in contrib extensions (namely `5`)
- successive core versions predominant over older ones (later is better)

It doesn't make sense that the major version of the mixin is tied to the core major version. If we are slightly stuck with 5 for the major version now, and want to beat all the versions already out there up to `5.82`, would switching the core version to `5.100.[core-version]` work?

colemanw commented:
> But in practice, is civimix-schema copied around?

That's a question Tim & I couldn't agree on when designing Entity-Framework v2. The mixin only needs to get copied into an extension if the author *does* want to use EFv2, but does *not* want to bump up their core version requirement to something recent-ish.
I felt that scenario was not worth the trouble of supporting; Tim felt differently and so he went ahead and created this mixin-based system that allows extensions to essentially backport their own copy of EFv2 in the form of `civimix-schema`.
In practice, has anyone actually used it? I'd guess probably not. However, there's another advantage to the mixin-based approach that Tim designed, which is that if Core ever gets crazy and implements quantum-based entities or something like that and calls it EFv3, extensions can comfortably remain on EFv2 with their own mixins.

totten commented:
At this point, the only failure is the ` api\v4\SearchDisplay\SearchRunTest.testRunWithTags`

ufundo commented:
Looks good @totten - thanks!

ufundo commented:
![image](https://github.com/user-attachments/assets/6b999b43-e50c-4046-9da7-4bbb305b72a6)

:rocket:
